### PR TITLE
feat: add AMD ROCm/HIP support (MI300X/MI308X/MI325X)

### PR DESCRIPTION
**Pull request by andyluo7**
## Summary

Add AMD ROCm support to autoresearch, enabling training on AMD Instinct GPUs (MI300X, MI308X, MI325X, MI250X).

All changes are in `train.py` only — `prepare.py` and the model architecture are unchanged.

### 1. Flash Attention 3 → PyTorch SDPA on ROCm
- FA3 (`kernels` package) is NVIDIA-only
- On ROCm, use `F.scaled_dot_product_attention` which dispatches to **AOTriton** automatically
- **Trade-off:** SDPA does not support `window_size`, so the SSSL sliding-window pattern degrades to full causal attention on all layers

### 2. Conditional `torch.compile`
- PyTorch ≤2.8 on ROCm has a strict dtype check in `lerp_()` that breaks inside compiled graphs (bfloat16 tensors mixed with float32 scalar weights)
- `torch.compile` is disabled on ROCm; re-enable with PyTorch 2.9+ which fixes this natively
- Applies to: model compilation, `adamw_step_fused`, and `muon_step_fused`

### 3. `lerp_` dtype casting
- Explicit dtype casting for `lerp_` weight argument to handle ROCm's stricter type checking in eager mode

### 4. Auto-detect GPU peak FLOPS
- Added lookup table covering H100, H200, A100, B200 (NVIDIA) and MI300X, MI308X, MI325X, MI250X (AMD)
- Enables accurate MFU reporting on any supported GPU
- Falls back to H100 for unknown GPUs with a warning

## Results on AMD MI308X

| Metric | Value |
|---|---|
| **val_bpb** | **1.520835** |
| Training time | 302.2s |
| Peak VRAM | ~97 GB / 192 GB |
| MFU | 3.18% (eager mode, no torch.compile) |
| Throughput | ~161K tok/sec |
| Total tokens | 54.5M |
| Steps | 104 |

> **Note:** MFU is low because `torch.compile` is disabled on PyTorch 2.6.0/ROCm. With PyTorch 2.9+ on ROCm (which fixes `lerp_` and compilation), MFU is expected to reach 15-30%+.

## How to Run on AMD

```bash
# On a node with AMD MI300X/MI308X + ROCm drivers:
git clone https://github.com/karpathy/autoresearch.git
cd autoresearch

docker run --rm \
  --device=/dev/kfd --device=/dev/dri \
  --group-add video --shm-size=64g --ipc=host \
  -v $(pwd):/workspace/autoresearch \
  -v ~/.cache/autoresearch:/root/.cache/autoresearch \
  -e HIP_VISIBLE_DEVICES=0 \
  -w /workspace/autoresearch \
  rocm/pytorch:rocm6.4.3_ubuntu24.04_py3.12_pytorch_release_2.6.0 \
  bash -c "pip install matplotlib pyarrow rustbpe tiktoken && \
           python3 prepare.py --num-shards 10 && \
           python3 train.py"
```

## Reference
- Blog post with detailed walkthrough: https://zjli2013.github.io/#/2026/autoresearch-rocm-mi308x

## Backward Compatibility
- Zero changes to NVIDIA/CUDA codepath — `IS_ROCM` flag gates all ROCm-specific logic
- `prepare.py` unchanged
- Model architecture unchanged

### CODE REVIEW

**Comment by huanrwan-amd** [comment]
Hi @andyluo7, thank you for support with AMD GPU. Have you tried with ROCm 7 image? rocm/pytorch:rocm7.1.1_ubuntu22.04_py3.11_pytorch_release_2.10.0 

need one line change for ROCm 7
```
  --- a/train.py
  +++ b/train.py
  @@ -358,7 +358,7 @@ def muon_step_fused(stacked_grads, stacked_params, momentum_buffer, second_momen
       red_dim_size = g.size(red_dim)
       v_norm_sq = v_mean.sum(dim=(-2, -1), keepdim=True) * red_dim_size
       v_norm = v_norm_sq.sqrt()
  -    second_momentum_buffer.lerp_(v_mean.to(dtype=second_momentum_buffer.dtype), 1 - beta2)
  +    second_momentum_buffer.lerp_(v_mean.to(dtype=second_momentum_buffer.dtype), (1 - beta2).to(dtype=second_momentum_buffer.dtype))
       step_size = second_momentum_buffer.clamp_min(1e-10).rsqrt()
       scaled_sq_sum = (v_mean * red_dim_size) * step_size.float().square()
       v_norm_new = scaled_sq_sum.sum(dim=(-2, -1), keepdim=True).sqrt()

```

**Comment by andyluo7** [comment]
Thanks @huanrwan-amd! Great catch. I've applied your fix for the `muon_step_fused` `lerp_` dtype casting — pushed in commit bcd28da.

This should now work on both ROCm 6.4.3 (PyTorch 2.6) and ROCm 7.1.1 (PyTorch 2.10). 

Have you been able to re-enable `torch.compile` on ROCm 7? That would be the big MFU unlock.

**Comment by huanrwan-amd** [comment]
@andyluo7 Thanks!
I have not re-enable torch.compile on ROCm 7 yet. That could help us on the performance side. Would be our next step.

**Comment by andyluo7** [comment]
## Test Results — Verified on AMD MI300X (single GPU)

Both ROCm versions pass ✅

| Metric | ROCm 6.4.3 / PyTorch 2.6.0 | ROCm 7.1.1 / PyTorch 2.9.1 |
|---|---|---|
| **val_bpb** | **1.160** | **1.121** |
| MFU | 7.48% | 10.36% |
| Throughput | ~415K tok/sec | ~567K tok/sec |
| Steps (300s budget) | 244 | 335 |
| Total tokens | 127.9M | 175.6M |
| Peak VRAM | 97.0 GB | 88.5 GB |
| Final loss | 3.27 | 3.15 |

### Key Observations

1. **ROCm 7.1.1 is 37% faster** — likely due to AOTriton/HIP runtime improvements
2. **9% less VRAM on 7.1.1** — better memory management in newer PyTorch
3. `torch.compile` still disabled on both — enabling on 7.1.1 is the next MFU unlock
4. All code changes work correctly on both versions
5. @huanrwan-amd's `lerp_` fix (commit bcd28da) is essential for 7.1.1

### Docker Commands Used

```bash
# ROCm 6.4.3
docker run --rm --network=host --device=/dev/kfd --device=/dev/dri \
  --group-add video --shm-size=64g --ipc=host \
  -e HIP_VISIBLE_DEVICES=0 \
  rocm/pytorch:rocm6.4.3_ubuntu24.04_py3.12_pytorch_release_2.6.0

# ROCm 7.1.1
docker run --rm --network=host --device=/dev/kfd --device=/dev/dri \
  --group-add video --shm-size=64g --ipc=host \
  -e HIP_VISIBLE_DEVICES=1 \
  rocm/pytorch:rocm7.1.1_ubuntu24.04_py3.12_pytorch_release_2.9.1
```
